### PR TITLE
Added exception when the dns request is for a portal detection url

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Added an exception when the dns request concern a fqdn for portal detection mechanism.

# Impacts
DNS

# Issue
- When a google oauth authentication source the domain *.gstatic.com is configured for dns bypass but the portal detection fqdn is connectivitycheck.gstatic.com and pfdns return the real ip address instead of the portal one.

# Delete branch after merge
YES

# Checklist
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries

## Enhancements
* Exception for portal detecion url in pfdns

### CODE REVIEW

**Comment by julsemaan** [comment]
This will need a rebase

**Comment by fdurand** [comment]
Rebased done

File: go/coredns/plugin/pfdns/pfdns.go [hunk 1]
```diff
@@ -29,27 +30,29 @@ import (
...
+	FqdnDomainPort      map[*regexp.Regexp][]string
+	Network             map[*net.IPNet]net.IP
+	NetworkType         map[*net.IPNet]string
+	DNSFilter           *cache.Cache
+	IpsetCache          *cache.Cache
+	apiClient           *unifiedapiclient.Client
+	refreshLauncher     *sync.Once
+	PortalFQDN          map[int]map[*net.IPNet]*regexp.Regexp
+	mutex               sync.Mutex
+	detectionmechanisms []*regexp.Regexp
```
Comment (julsemaan) [inline go/coredns/plugin/pfdns/pfdns.go]: Use camelcase like the other variable names

File: go/coredns/plugin/pfdns/pfdns.go [hunk 2]
```diff
@@ -699,3 +713,42 @@ func (pf *pfdns) PortalFQDNInit(ctx context.Context) error {
...
 	return mac + category + strconv.FormatBool(security_event) + qname
 }
+
+func (pf *pfdns) MakeDetectionMecanism() error {
+	portal := pfconfigdriver.Config.PfConf.CaptivePortal
+	pf.detectionmechanisms = make([]*regexp.Regexp, 0)
+	var err error
+	for _, v := range portal.DetectionMecanismUrls {
+		fqdn, err := url.Parse(v)
+		if err != nil {
```
Comment (julsemaan) [inline go/coredns/plugin/pfdns/pfdns.go:723]: Log an error ?

File: go/coredns/plugin/pfdns/pfdns.go [hunk 3]
```diff
@@ -699,3 +713,42 @@ func (pf *pfdns) PortalFQDNInit(ctx context.Context) error {
...
+			continue
+		}
+		err = pf.addDetectionMechanismsToList(fqdn.Host)
+	}
+	return err
+}
+
+// addDetectionMechanismsToList add all detection mechanisms in a list
+func (pf *pfdns) addDetectionMechanismsToList(r string) error {
+	rgx, err := regexp.Compile(r)
```
Comment (julsemaan) [inline go/coredns/plugin/pfdns/pfdns.go:734]: Log error ?

**Comment by fdurand** [comment]
Bump

**Comment by extrafu** [comment]
@julsemaan Please review and merge

**Pull request merged**